kernel/USER_GUIDE.md

```diff
@@ -0,0 +1,604 @@
+
+
+#### Implement JsonHandler
+This client interface allows connector to use plug-in their own JSON handing code and expose it to the Delta Kernel.
+
+##### Method contextualizeFileReads() 
+
+This method associates the information of a to-be-read file to a connector-specific context object, called `FileReadContext`. Implementations can use this method to inject connector-specific information to optimize the file read.
+
+**Requirements and guarantees: **Any implementation must adhere to the following guarantees.
+
+* The output iterator must maintain the same ordering as the input iterator. For example, if file1 is before file2 in the input iterator, then all the contexts associated with file1 must be before those of file2 in the output iterator.
+
+
+##### Method readJsonFiles() 
+
+This method takes as input information as `FileReadContext`s and returns the data in a series of columnar batches. The columns to be read from the JSON file are defined by the physical schema, and the return batches must match that schema. To implement this method, you may have to first implement your own `<code>[ColumnarBatch](https://delta-io.github.io/delta/snapshot/kernel-api/java/api//io/delta/kernel/data/ColumnarBatch.html)`</code> and `<code>[ColumnVector](https://delta-io.github.io/delta/snapshot/kernel-api/java/api//io/delta/kernel/data/ColumnVector.html)`</code> which is used to represent the in-memory data generated from the JSON files.
+
+When identifying the columns to read, note that there are multiple types of columns in the physical schema (represented as a StructType). 
+
+#### Method parseJson
+
+This method allows parsing a `ColumnVector` of JSON string into the output format specified by the `outputSchema`. If a given column column `outputSchema` is not found, then a null value is returned.
+
+
+#### Implement ColumnarBatch and ColumnVector 
+
+ColumnarBatch and ColumnVector two interfaces represent the data read into memory from files. This representation can have a significant impact on query performance. Each engine likely has a native representation of in-memory data with which it applies data transformations operations. For example, in Spark, the row data is internally represented as `UnsafeRow` for efficient processing. So it's best to read the Parquet file data directly into vectors and batches of the native format to avoid potentially costly in-memory data format conversions. So the recommended approach is to build wrapper classes that extend the two interfaces but internally use engine-native classes to store the data. When the connector has to forward the columnar batches received from the kernel to the engine, it has to be smart enough to skip converting vectors and batches that are already in the engine-native format. This will be discussed in more detail later. 
```

Comment:
One sentence per line is easier to edit and diff over time...
```suggestion
`ColumnarBatch` and `ColumnVector` two interfaces represent the data read into memory from files. 
This representation can have a significant impact on query performance. 
Each engine likely has a native representation of in-memory data with which it applies data transformations operations. 
For example, in Spark, the row data is internally represented as `UnsafeRow` for efficient processing. 
So it's best to read the Parquet file data directly into vectors and batches of the native format to avoid potentially costly in-memory data format conversions. 
The recommended approach is to build wrapper classes that extend the two interfaces but internally use engine-native classes to store the data. 
When the connector has to forward the columnar batches received from the kernel to the engine, it has to be smart enough to skip converting vectors and batches that are already in the engine-native format. 
This will be discussed in more detail later. 
```